xgcm/test/test_integrate.py

```diff
@@ -0,0 +1,355 @@
+    }
+
+
+###################
+# Test metric internals (this should be moved I suppose)
+##################
+@pytest.mark.parametrize(
+    "key, metric_vars",
+    [
+        # (("X"), ["dx_t"]), # this fails, can we raise a more useful error?
+        (("X",), ["dx_t"]),
+        (("X",), []),  # Should this pass?
+        (("X",), "dx_t"),
+        (("X", "Y"), ["area_t"]),
+        (
+            ("X", "Y"),
+            ["area_t", "area_t"],
+        ),  # this should also trigger an error, it does not
+        (("X", "Y"), ["area_t", "area_e", "area_n", "area_ne"]),
+        (("X", "Y", "Z"), ["volume_t"]),
+    ],
+)
+def test_assign_get_metric(key, metric_vars):
+    ds_full = datasets()
+    ds = ds_full["C"]
+    grid = Grid(
+        ds,
+        coords=ds_full["coords"],
+        # periodic=True,
+    )
+    grid._assign_metrics({key: metric_vars})
+
+    # convert key
+    key = frozenset(key)
+
+    # if input is str convert to list
+    if isinstance(metric_vars, str):
+        metric_vars = [metric_vars]
+    expected = [ds[mv].reset_coords(drop=True) for mv in metric_vars]
+
+    assert all([a.equals(b) for a, b in zip(grid._metrics[key], expected)])
+
+
+@pytest.mark.parametrize(
+    "axes, data_var, drop_vars, metric_expected_list, expected_error",
+    [
+        ("X", "tracer", None, ["dx_t"], None),
+        (["X", "Y"], "tracer", None, ["area_t"], None),
+        (
+            ("X", "Y"),
+            "tracer",
+            None,
+            ["area_t"],
+            None,
+        ),  # should we be able to pass a tuple aswell as a list?
+        (["X", "Y", "Z"], "tracer", None, ["volume_t"], None),
+        (["X"], "u", None, ["dx_e"], None),
+        (["X", "Y"], "u", None, ["area_e"], None),
+        (("X", "Y", "Z"), "u", None, ["volume_t"], KeyError),  # This should error out
+        # reconstructed cases
+        (["X", "Y"], "tracer", ["area_t"], ["dx_t", "dy_t"], None),
+        (["X", "Y", "Z"], "tracer", ["volume_t"], ["area_t", "dz_t"], None),
+    ],
+)
+def test_get_metric(axes, data_var, drop_vars, metric_expected_list, expected_error):
+    ds_full = datasets()
+    ds = ds_full["C"]
+    metrics = ds_full["metrics"]
+    # drop metrics according to drop_vars input, and remove from metrics input
+    if drop_vars:
+        print(drop_vars)
+        ds = ds.drop(drop_vars)
+        metrics = {k: [a for a in v if a not in drop_vars] for k, v in metrics.items()}
+
+    grid = Grid(
+        ds,
+        coords=ds_full["coords"],
+        metrics=metrics,
+        # periodic=True,
+    )
+    if expected_error:
+        with pytest.raises(expected_error):
+            metric = grid.get_metric(ds[data_var], axes)
+    else:
+        metric = grid.get_metric(ds[data_var], axes)
+        expected_metrics = [
+            ds[me].reset_coords(drop=True) for me in metric_expected_list
+        ]
+        expected = functools.reduce(operator.mul, expected_metrics, 1)
+        assert metric.equals(expected)
+
```

Comment:
Up until here stuff should probably go to #136

Reply:
This is now very redundant with the dataset generator in test_metrics.

Any chance you could factor these out into a standalone module in the tests directory?